sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java

```diff
@@ -0,0 +1,1071 @@
+     */
+    public <K, DataT, InputT, OutputT> ComposedKeyedCombineFn<DataT, K> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFn<InputT, ?, OutputT> combineFn,
+        TupleTag<OutputT> outputTag) {
+      return with(extractInputFn, combineFn.<K>asKeyedFn(), outputTag);
+    }
+
+    /**
+     * Returns a {@link ComposedKeyedCombineFnWithContext} that can take additional
+     * {@link PerKeyCombineFn PerKeyCombineFns} and apply them as a single combine function.
+     */
+    public <K, DataT, InputT, OutputT> ComposedKeyedCombineFnWithContext<DataT, K> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFnWithContext<InputT, ?, OutputT> combineFnWithContext,
+        TupleTag<OutputT> outputTag) {
+      return with(extractInputFn, combineFnWithContext.<K>asKeyedFn(), outputTag);
+    }
+  }
+
+  /**
+   * A builder class to construct a composed {@link GlobalCombineFn}.
+   */
+  public static class ComposeCombineFnBuilder {
+    /**
+     * Returns a {@link ComposedCombineFn} that can take additional
+     * {@link GlobalCombineFn GlobalCombineFns} and apply them as a single combine function.
+     *
+     * <p>The {@link ComposedCombineFn} extracts inputs from {@code DataT} with
+     * the {@code extractInputFn} and combines them with the {@code combineFn},
+     * and then it outputs each combined value with a {@link TupleTag} to a
+     * {@link CoCombineResult}.
+     */
+    public <DataT, InputT, OutputT> ComposedCombineFn<DataT> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFn<InputT, ?, OutputT> combineFn,
+        TupleTag<OutputT> outputTag) {
+      return new ComposedCombineFn<DataT>()
+          .with(extractInputFn, combineFn, outputTag);
+    }
+
+    /**
+     * Returns a {@link ComposedCombineFnWithContext} that can take additional
+     * {@link GlobalCombineFn GlobalCombineFns} and apply them as a single combine function.
+     *
+     * <p>The {@link ComposedCombineFnWithContext} extracts inputs from {@code DataT} with
+     * the {@code extractInputFn} and combines them with the {@code combineFnWithContext},
+     * and then it outputs each combined value with a {@link TupleTag} to a
+     * {@link CoCombineResult}.
+     */
+    public <DataT, InputT, OutputT> ComposedCombineFnWithContext<DataT> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFnWithContext<InputT, ?, OutputT> combineFnWithContext,
+        TupleTag<OutputT> outputTag) {
+      return new ComposedCombineFnWithContext<DataT>()
+          .with(extractInputFn, combineFnWithContext, outputTag);
+    }
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * A tuple of {@link Object Objects} produced for a given key from the composed combine functions.
+   */
+  public static class CoCombineResult implements Serializable {
```

Comment:
If we have a combine sub-package, I would put CoCombineResult as a top-level class there.
But, we don't have that now, and I don't find other better locations.

For example, I don't like to put it as a top-level class in transforms, since it is not the data type from any Transforms.
And, I think it fit better in here, because CoCombineResult is the output of combine functions. And, it ties with how composed combine functions are implemented.